src/symbols_linux.cpp

```diff
@@ -804,13 +817,20 @@ void Symbols::parseLibraries(CodeCacheArray* array, bool kernel_symbols) {
 
             // Protect library from unloading while parsing in-memory ELF program headers.
             // Also, dlopen() ensures the library is fully loaded.
-            // Main executable and ld-linux interpreter cannot be dlopen'ed, but dlerror() returns NULL for them.
+            // Main executable and ld-linux interpreter cannot be dlopen'ed, but dlerror() "in some cases" returns NULL for them.
             void* handle = dlopen(lib.file, RTLD_LAZY | RTLD_NOLOAD);
-            if (handle != NULL || dlerror() == NULL || OS::isMusl()) {
+            const char* dlerror_str = dlerror();
+
+            // Main executable will return NULL from handle, we need to parse it anyway.
+            // dlopen is not required in this case as the main exe cannot be unloaded.
+            bool isMainExe = handle == NULL && exePath != NULL && strcmp(lib.file, exePath) == 0;
+            bool dlopenSuccess = handle != NULL || dlerror_str == NULL || OS::isMusl();
```

Comment:
Maybe we don't need `OS::isMusl()` anymore? The only situation when this was appropriate is for the main executable, in which case `isMainExe` is `true` anyways